core/multigrid/amgx_pgm.cpp

```diff
@@ -0,0 +1,173 @@
+GKO_REGISTER_OPERATION(count_unagg, amgx_pgm::count_unagg);
+GKO_REGISTER_OPERATION(renumber, amgx_pgm::renumber);
+GKO_REGISTER_OPERATION(extract_diag, amgx_pgm::extract_diag);
+GKO_REGISTER_OPERATION(find_strongest_neighbor,
+                       amgx_pgm::find_strongest_neighbor);
+GKO_REGISTER_OPERATION(assign_to_exist_agg, amgx_pgm::assign_to_exist_agg);
+GKO_REGISTER_OPERATION(amgx_pgm_generate, amgx_pgm::amgx_pgm_generate);
+
+
+}  // namespace amgx_pgm
+
+
+namespace {
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> amgx_pgm_generate(
+    std::shared_ptr<const Executor> exec,
+    const matrix::Csr<ValueType, IndexType> *source, const size_type num_agg,
+    const Array<IndexType> &agg)
+{
+    auto coarse = matrix::Csr<ValueType, IndexType>::create(
+        exec, dim<2>{num_agg, num_agg}, 0, source->get_strategy());
+    exec->run(amgx_pgm::make_amgx_pgm_generate(source, agg, coarse.get()));
+    matrix::CsrBuilder<ValueType, IndexType> builder(coarse.get());
+    return std::move(coarse);
+}
+
+
+}  // namespace
+
+
+template <typename ValueType, typename IndexType>
+void AmgxPgm<ValueType, IndexType>::generate()
+{
+    using matrix_type = matrix::Csr<ValueType, IndexType>;
+    auto exec = this->get_executor();
+    const auto num = this->system_matrix_->get_size()[0];
+    Array<ValueType> diag(this->get_executor(), num);
+    Array<IndexType> strongest_neighbor(this->get_executor(), num);
+    Array<IndexType> intermediate_agg(this->get_executor(),
+                                      parameters_.deterministic * num);
+    const auto amgxpgm_op = gko::as<matrix_type>(this->system_matrix_.get());
+    // Initial agg = -1
+    exec->run(amgx_pgm::make_initial(agg_));
+    // Extract the diagonal value of matrix
+    exec->run(amgx_pgm::make_extract_diag(amgxpgm_op, diag));
+    size_type num_unagg{0};
+    size_type num_unagg_prev{0};
+    for (int i = 0; i < parameters_.max_iterations; i++) {
+        // Find the strongest neighbor of each row
+        exec->run(amgx_pgm::make_find_strongest_neighbor(amgxpgm_op, diag, agg_,
+                                                         strongest_neighbor));
+        // Match edges
+        exec->run(amgx_pgm::make_match_edge(strongest_neighbor, agg_));
+        // Get the num_unagg
+        exec->run(amgx_pgm::make_count_unagg(agg_, &num_unagg));
+        // no new match, all match, or the ratio of num_unagg/num is lower
+        // than parameter.max_unassigned_percentage
+        if (num_unagg == 0 || num_unagg == num_unagg_prev ||
+            static_cast<double>(num_unagg) / num <
+                parameters_.max_unassigned_percentage) {
+            break;
+        }
+        num_unagg_prev = num_unagg;
+    }
+    // Handle the left unassign points
+    if (num_unagg != 0 && intermediate_agg.get_num_elems() > 0) {
+        // copy the agg to intermediate_agg
+        intermediate_agg = agg_;
+    }
+    while (num_unagg != 0) {
```

Comment:
Can it happen that we need to call assign_to_exist_agg multiple times? To me, it looks like every unassigned row will be assigned after the kernel finishes.

Reply:
You are right. it will assign the unagg to one agg group or itself.